File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<vector<int>>ans;
public:
    void solve(TreeNode* root, vector<int>v, int sum, int &t){
        if(root == NULL) return;
        v.push_back(root->val);
        sum += root->val;
        if(root->left == NULL && root->right == NULL){
            if(sum == t)    ans.push_back(v);
        }
        solve(root->left, v, sum, t);
        solve(root->right, v, sum, t);
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        if(root == NULL) return {};
        solve(root, {}, 0, targetSum);
        return ans;
    }
};

File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>>ans;
    void helper(TreeNode root, int targetSum, List<Integer>list){
        if(root == null) return;
        if(root.left == null && root.right==null && targetSum-root.val==0){
            list.add(root.val);
            ans.add(new ArrayList(list));
            list.remove(list.size()-1);
            return;
        }
        list.add(root.val);
        helper(root.left, targetSum-root.val, list);
        helper(root.right, targetSum-root.val, list);
        if(list.size()>0) list.remove(list.size()-1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        ans = new ArrayList<>();
        if(root == null) return ans;
        helper(root, targetSum, new ArrayList());
        return ans;
    }
}